tests/0105-regex_rebalance.c

```diff
@@ -0,0 +1,141 @@
+/*
+ * librdkafka - Apache Kafka C library
+ *
+ * Copyright (c) 2012-2015, Magnus Edenhill
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "test.h"
+
+/**
+ * Issue #2121:
+ * Test handling of asymmetric subscription race condition; if two consumers
+ * subscribe to "^.*", a new topic may not have been subscribed
+ * by both at the time rebalance is triggered.
+ */
+
+#include "rdkafka.h"
+
+
+static rd_kafka_t *c1, *c2;
+static rd_kafka_resp_err_t state1, state2;
+
+static void rebalance_cb (rd_kafka_t *rk, rd_kafka_resp_err_t err,
+                          rd_kafka_topic_partition_list_t *parts, void *opaque) {
+        rd_kafka_resp_err_t *statep = NULL;
+
+        if (rk == c1)
+                statep = &state1;
+        else if (rk == c2)
+                statep = &state2;
+        else
+                TEST_FAIL("Invalid rk %p", rk);
+
+        TEST_SAY("Rebalance for %s: %s:\n", rd_kafka_name(rk), rd_kafka_err2str(err));
+        test_print_partition_list(parts);
+
+        if (err == RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS)
+                rd_kafka_assign(rk, parts);
+        else if (err == RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS)
+                rd_kafka_assign(rk, NULL);
+
+        *statep = err;
+}
+
+static void test_regex_rebalance(const char* assignor) {
+        const char *topic = test_mk_topic_name(__FUNCTION__ + 5, 1);
+        int64_t ts_start;
+        int wait_sec;
+        state1 = state2 = 0;
+        char newtopic[50];
+        strcpy(newtopic, topic);
```

Comment:
this is unsafe, you dont know the size of the topic.
Use rd_snprintf and `newtopic[256]`